### PR TITLE
Feature/homework 5

### PR DESCRIPTION
**Pull request by s0557917**
- Implementation of homework 5
- We left this <a href="https://github.com/Systems-Development-and-Frameworks/abteilung6/pulls#pullrequestreview-561846587"> pull review </a>

### CODE REVIEW

File: backend/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# TheCodeNinjas - Backend
+
+## GraphQL Architecture
+When having to choose between a locally hosted GraphQL implementation (neo4j-graphql-js library) and a Remote GraphQL API (headless GraphCMS API), we chose the headless variant as, firstly, it allowed us to work on a central database without the need of setting it up multiple times and having to invest effort and time into maintaining it.
+We also chose it, as it was an interesting API and we wanted to learn how to work with it.
```
Comment (Marcelloloco) [inline backend/README.md:5]: Motivation objective done! :+1:

File: backend/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+## Installation
```
Comment (Marcelloloco) [inline backend/README.md:8]: Installation manual objective done! :+1:

File: backend/.env.example [hunk 3]
```diff
@@ -0,0 +1,6 @@
+JWT_SECRET=
+JWT_EXPIRES_IN=
+PORT=
+SALT_ROUNDS=
+GRAPH_CMS_ENDPOINT=
+GRAPH_CMS_API_TOKEN=
```
Comment (Marcelloloco) [inline backend/.env.example:6]: No API key committed. :+1:

File: backend/src/create-permissions.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        if (post) {
+          return post.author.id === jwtPayload.id;
+        }
+      }
+    }
+
+    return false;
+  });
+
+  return shield(
```
Comment (Marcelloloco) [inline backend/src/create-permissions.ts:52]: Permissions handled for calling mutations and queries. :+1:

File: backend/src/create-resolvers.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    Query: {
+      ...postsQuery(subSchemas, executor),
+      ...personsQuery(subSchemas, executor),
+    },
+    Mutation: {
+      ...personsMutation(subSchemas, executor),
+      ...postsMutation(subSchemas, executor),
+    },
+    Post: postsProperties(subSchemas, executor),
+    Person: personsProperties(subSchemas, executor),
```
Comment (Marcelloloco) [inline backend/src/create-resolvers.ts:27]: Resolvers delegation for persistent communication with Database :+1:

File: backend/src/resolvers/persons.resolver.spec.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+      );
+
+      expect(response).toMatchObject({
+        data: {
+          signup: expect.any(String),
+        },
+      });
+    });
+  });
+});
```
Comment (Marcelloloco) [inline backend/src/resolvers/persons.resolver.spec.ts:145]: Tests adjusted and are running fine on Github actions :+1:

**Review by Marcelloloco — APPROVED** [review]
Good job! 

You also requested a review, so there's another :+1:

File: backend/src/create-permissions.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        signup: allow,
+        downvote: isAuthenticated,
+        upvote: isAuthenticated,
+        write: isAuthenticated,
+        delete: isOwner,
+      },
+    },
+    {
+      allowExternalErrors: true,
+      fallbackRule: allow,
```
Comment (medizinmensch) [inline backend/src/create-permissions.ts:84]: Here, as well as in line 55 and 73, you set your fallback rule as "allow". This default gives access to all further mutations and queries, especially to the ones in your subschema. This means you're getting an :x: here:

:star: You deny the client to   call mutations and queries of your subschema directly.

**Review by medizinmensch — CHANGES_REQUESTED** [review]
Hi Team!

Thanks for your submission!
Your code looks code, even though it's a bit excessive (you have more than twice as many changed files as other groups), but I guess this is because you make use of TypeScript features like interfaces.

The tests run through and also successfully fail if I change the `expected` values. 👍 

Sadly, the server itself wont start for me. I get this error:

```
$ npm run start
> backend@1.0.0 start /home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend
> ts-node src/index.ts

Error: It seems like you have applied rules to Query.persons fields but Shield cannot find them in your schema.
    at applyRuleToType (/home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-shield/src/generator.ts:151:13)
    at /home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-shield/src/generator.ts:265:25
    at Array.reduce (<anonymous>)
    at generateMiddlewareFromSchemaAndRuleTree (/home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-shield/src/generator.ts:259:8)
    at MiddlewareGenerator.generator (/home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-shield/src/generator.ts:293:5)
    at MiddlewareGenerator.generate (/home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-middleware/src/generator.ts:19:17)
    at /home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-middleware/src/middleware.ts:72:25
    at Array.map (<anonymous>)
    at applyMiddlewareWithOptions (/home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-middleware/src/middleware.ts:70:45)
    at Object.applyMiddleware (/home/youri/dev/uni/sdaf/groups/TheCodeNinjas/backend/node_modules/graphql-middleware/src/middleware.ts:129:10)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! backend@1.0.0 start: `ts-node src/index.ts`
npm ERR! Exit status 1
```

I could not quickly make out the issue. I'm on the `/feature/homework-5` branch, pulled properly and didn't change any files (except the `.env`). I added one line with `console.log(fetchResult.json()` to `create-graph-cms-schema.ts`/`executor` to confirm a working connection and it looks like my GraphCMS settings are correct. My GraphCMS Schema also looks like yours:

Photoshop_5EC5JRKmuI


I don't know what the issue is. Do you guys have any idea? Until this is fixed, I can't give you the points for:
> :star:   :star: Data created in mutations are persisted and your queries and mutations   still work as expected.

Below, you'll find a comment on why I have to deduct one :star:. 

Also, you commited your JWT_Secret in your `README` file, is that correct. I have to deduct another point for that :x:
> :star: For not committing   sensitive secrets (e.g. API keys) unencrypted to the repository.

So far, you get **6/10** :star: for the following:

:star: For choosing a scenario   and writing installation instructions in README.md.
:star:   For explaining WHY you have chosen the scenario in the README.md.
:star: :star: Your software   tests are free of side effects.
:star:   Your database is never left in an invalid state.
:star:   For requesting a review and reviewing another team's PR.

If we get the code running and find out it was my fault, I can give you 2 more :star:.



Youri

File: backend/src/create-permissions.ts (hunk 7 above)
Comment (s0557917) [reply backend/src/create-permissions.ts:84]: Hey Youri, for some reason I can only reply to this comment and not the whole PR Review.
Regarding the server not starting: I could not reproduce the error you are getting. I cloned the project and added our values to the .env file in the backend folder and it works. Your GraphCMS schema also looks correct so I am not sure where the error happens.

**Review by medizinmensch — APPROVED** [review]
We talked today, I will re-review your code tomorrow. Until then, you can already merge :)

**Pull request merged**